core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java

```diff
@@ -319,6 +402,1345 @@ public void testNamespaceWithDot() {
+    PartitionSpec expected = update.apply();
+
+    // update the schema concurrently so that the first update fails, but can succeed on retry
+    catalog.loadTable(TABLE).updateSchema()
+        .addColumn("another_col", Types.StringType.get())
+        .commit();
+
+    // commit the original update
+    update.commit();
+
+    Table loaded = catalog.loadTable(TABLE);
+
+    // the spec ID may not match, so check equality of the fields
+    Assert.assertEquals("Loaded table should have expected spec", expected.fields(), loaded.spec().fields());
+  }
+
+  @Test
+  public void testUpdateTableSpecConflict() {
+    C catalog = catalog();
+
+    Table table = catalog.buildTable(TABLE, SCHEMA).withPartitionSpec(SPEC).create();
+
+    UpdatePartitionSpec update = table.updateSpec()
+        .addField("shard", Expressions.bucket("data", 16));
+
+    // update the spec concurrently so that the original update fails
+    UpdatePartitionSpec concurrent = catalog.loadTable(TABLE).updateSpec()
+        .removeField(Expressions.bucket("id", 16));
+    PartitionSpec expected = concurrent.apply();
+    concurrent.commit();
+
+    // attempt to commit the original update
+    String expectedMessage = supportsServerSideRetry() ?
+        "Requirement failed: default partition spec changed" : "Cannot commit";
+    AssertHelpers.assertThrows("Second partition spec update commit should fail because of a conflict",
+        CommitFailedException.class, expectedMessage, update::commit);
+
+    Table loaded = catalog.loadTable(TABLE);
+
+    // the spec ID may not match, so check equality of the fields
+    Assert.assertEquals("Loaded table should have expected spec", expected.fields(), loaded.spec().fields());
+  }
+
+  @Test
+  public void testUpdateTableAssignmentSpecConflict() {
+    C catalog = catalog();
+
+    Table table = catalog.buildTable(TABLE, SCHEMA).create();
+
+    UpdatePartitionSpec update = table.updateSpec()
+        .addField("shard", Expressions.bucket("id", 16));
+
+    // update the spec concurrently so that the original update fails
+    UpdatePartitionSpec concurrent = catalog.loadTable(TABLE).updateSpec()
+        .addField("shard", Expressions.truncate("id", 100));
+    PartitionSpec expected = concurrent.apply();
+    concurrent.commit();
+
+    // attempt to commit the original update
+    String expectedMessage = supportsServerSideRetry() ?
+        "Requirement failed: last assigned partition id changed" : "Cannot commit";
+    AssertHelpers.assertThrows("Second partition spec update commit should fail because of a conflict",
+        CommitFailedException.class, expectedMessage, update::commit);
+
+    Table loaded = catalog.loadTable(TABLE);
+
+    // the spec ID may not match, so check equality of the fields
+    Assert.assertEquals("Loaded table should have expected spec", expected.fields(), loaded.spec().fields());
+  }
+
+  @Test
+  public void testUpdateTableSortOrder() {
```

Comment:
`testUpdateTableSortOrder` and `testUpdateTableSortOrderServerSideRetry` for example only differ in 1-2 lines, so I woder if we can improve the tests a bit and try and reduce code duplication (in a follow up PR)